### DISCUSSION TITLE
Add WASM support to the library

### DISCUSSION THREAD

**Discussion by aadilshabier** [gabhijit/scalpel#61] [Ideas]
WASM is a way to bring performant languages to the web, and provide alternatives for Javascript. Making the library WASM compatible will enable the library to be used in web applications and help widen adoption.
This can optionally be done by making the entire library `#![no_std]`.

**Comment by csking101** [comment]
One of the long-term goals is to be able to run scalpel on a browser, so there is no need for the browser to externally query and API endpoint to dissect the packets that it's receiving, it can do so directly internally.

**Comment by gabhijit** [comment]
One may look at something like following -
https://rustwasm.github.io/docs/book/reference/add-wasm-support-to-crate.html and see.
I quickly tried adding `wasm-unknown-emscripten` and `was-unknown-unknown` targets using `rustup target add <target-name>` and didn't face issues. But a more detailed analysis and a plan of action needs to be determined. First goal is to be able to define TODO tasks for that and then let's go after that.

Honestly my knowledge about whole WASM space is rather limited. :-)

**Reply by aadilshabier** [reply]
Yep, it seems like we might have good WASM support already! Our library doesn't do any of the things listed there. We need to try running tests now I think.

**Reply by gabhijit** [reply]
@aadilshabier : It might be a good idea to think of some demo code that can be used to demonstrate, `scalpel` can dissect packet in your browser. Something as simple as a hard coded JS data (hex dump of packets) with a very simple `index.html` -> where  you can `dissect` on click and the resulting packet is displayed as `json`. 

Also, it might make sense to add that target in the CI, once we have something working as a demo.

Anyone of you who might want to start with a draft PR for that is welcome.

**Reply by csking101** [reply]
Hi sir, could I take this up?

**Reply by gabhijit** [reply]
@csking101 : Sure !! May be a good idea is to start a draft PR referencing this discussion.

**Reply by csking101** [reply]
#64 references this discussion, I have made a simple demonstration of converting our crate to WASM and using the same to dissect packets in JS
